share/wake/lib/core/json.wake

```diff
@@ -169,3 +169,143 @@ export def x ==/ y = match x y
+export def normalizeJSONCompat: JSONNormalize =
+    def nfkcKeys =
+        map (editPairFirst unicodeIdentifier)
+    normalizeJSONIdentity
+    | setJSONNormalizeString (Pass _.unicodeCanonical)
+    | setJSONNormalizeDouble filterNonDigitJDouble
+    | setJSONNormalizeObject (deduplicateJObjectKeys _.nfkcKeys)
+
+# Fail on any `Double` values which can't be represented numerically.
+# Published JSON specifications describe their double values as allowing decimal
+# or exponential forms, but don't implement the full IEEE standard.
+def filterNonDigitJDouble (n: Double): Result Double Error = match n.dclass
+    DoubleInfinite =
+        failWithError "For compatibility, JSON doubles should not be infinite."
+    DoubleNaN =
+        failWithError "For compatibility, JSON doubles should not be NaN values."
+    _ =
+        Pass n
+
+# Attempt to simplify any `JObject`s which contain multiple instances of a key.
+# While the published specifications explicitly allow such duplicate keys, none
+# *require* implementations to accept them nor describes the behaviour by which
+# they should be handled; therefore, implementations differ in how they handle
+# duplicate keys -- if they do so at all -- and a generator aiming for
+# compatibility shouldn't output objects with duplicate keys.
+#
+# In particular, this function will concatenate the members of two `JArray`
+# values, treat `JNull` as a wildcard placeholder, and otherwise return the
+# original value if both instances are equal. It fails if a key is associated
+# with multiple values of different types or with different values of a type
+# which cannot be combined.
+def deduplicateJObjectKeys (dict: List (Pair String JValue)): Result (List (Pair String JValue)) Error =
+    def cmpKeysOnly (Pair k1 _) (Pair k2 _) =
+        k1 <* k2
+    def simplifyKeyGroups = match _
+        Nil =
+            # `simplifyKeyGroups` is only ever mapped over the output of
+            # `groupBy`, which is always a list of *non-empty* lists.
+            failWithError "groupBy invariant reached in deduplicateJObjectKeys"
+        (Pair key value), ps =
+            (value, map getPairSecond ps)
+            | rfoldr (simplifyValues key) JNull
+            | rmap (Pair key)
+    def simplifyValues key: JValue => JValue => Result JValue Error = match _ _
+        val JNull =
+            Pass val
+        JNull acc =
+            Pass acc
+        (JString val) (JString acc) if val ==* acc =
+            Pass (JString acc)
+        (JInteger val) (JInteger acc) if val == acc =
+            Pass (JInteger acc)
+        (JDouble val) (JDouble acc) if val ==. acc =
+            Pass (JDouble acc)
+        (JBoolean val) (JBoolean acc) if enor val acc =
+            Pass (JBoolean acc)
+        (JArray val) (JArray acc) =
+            Pass (JArray (val ++ acc))
+        (JObject val) (JObject acc) =
+            val ++ acc
+            | deduplicateJObjectKeys
+            | rmap JObject
```

Comment:
> The recursive call here is inappropriate. You should leave that to `normalizeJSON`, like the `JArray` case above does. ie: `Pass (JObject (val ++ acc))`

The restrictions are different.  The `JArray` case just takes the elements of one, sticks them on the elements of the other, and calls it done.  For `JObject`, both `val` and `acc` are guaranteed to not have any duplicate keys internally by this point, but the same key might occur in both and thus still needs to be filtered.  The alternate implementation would be to have `normalizeJSON` recurse *after* processing each level rather than before, but then you lose the ability to have it perform the heavy lifting for the above equality tests.

Reply:
Yes. I want `normalizeJSONWith` to be the one who does the recursion. It should work that way to ensure only a single pass processes any given JSON node. Furthermore, `normalizeJSONWith` has to be the one doing the recursion, because you want to apply the appropriate leaf normalizations in the descent.

Reply:
`normalizeJSONWith` definitely has to recurse for that exact reason, the question is just whether it processes top-down or bottom-up.  My argument for doing children before the current node (which does admittedly mean we can't always do just a single pass) is below.